File: clients/csharp/AmenoLink/Connection.cs
namespace AmenoLink;

internal class Connection : IConnection
{
    private const int MaxReconnectAttempts = 5;
    private const double ConnectionTimeoutSeconds = 5.0;

    private readonly HashSet<Action<ConnectionStatus>> listeners = [];
    private readonly Lock lockObject = new();
    private readonly ConnectionManager connectionManager;

    internal Connection(ConnectionManager connectionManager)
    {
        this.connectionManager = connectionManager;
    }

    public void Subscribe(Action<ConnectionStatus> listener)
    {
        lock (lockObject)
            listeners.Add(listener);
    }

    public void Unsubscribe(Action<ConnectionStatus> listener)
    {
        lock (lockObject)
            listeners.Remove(listener);
    }

    public void UnsubscribeAll()
    {
        lock (lockObject)
            listeners.Clear();
    }

    private void OnConnectionChanged(ConnectionStatus status)
    {
        List<Action<ConnectionStatus>> targets;
        lock (lockObject)
            targets = [.. listeners];

        foreach (var listener in targets)
            listener(status);
    }

    public Task Connect()
    {
        return connectionManager.Connect(
            onStatusChange: OnConnectionChanged,
            maxAttempts: MaxReconnectAttempts,
            timeoutSeconds: ConnectionTimeoutSeconds
        );
    }

    public Task Disconnect()
    {
        return connectionManager.Disconnect();
    }
}

File: clients/dart/amenolink/lib/src/connection.dart
import 'dart:async';
import 'connection_manager.dart';
import 'interfaces.dart' as i;

class Connection implements i.Connection {
  static const int _maxReconnectAttempts = 5;
  static const double _connectionTimeoutSeconds = 5.0;

  final Set<void Function(ConnectionStatus status)> _listeners = {};
  final ConnectionManager _connectionManager;

  Connection(this._connectionManager);

  @override
  void subscribe(void Function(ConnectionStatus status) listener) {
    _listeners.add(listener);
  }

  @override
  void unsubscribe(void Function(ConnectionStatus status) listener) {
    _listeners.remove(listener);
  }

  @override
  void unsubscribeAll() {
    _listeners.clear();
  }

  void _onConnectionChanged(ConnectionStatus status) {
    for (final listener in List<void Function(ConnectionStatus)>.from(_listeners)) {
      listener(status);
    }
  }

  @override
  Future<void> connect() {
    return _connectionManager.connect(
      onStatusChange: _onConnectionChanged,
      maxAttempts: _maxReconnectAttempts,
      timeoutSeconds: _connectionTimeoutSeconds,
    );
  }

  @override
  Future<void> disconnect() {
    return _connectionManager.disconnect();
  }
}

final connection = Connection(connectionManager);
